lib/std/simd.zig

```diff
@@ -0,0 +1,369 @@
+        const cycle = @divFloor(count_up, @splat(len, @intCast(i32, vecs_arr.len)));
+        const select_mask = repeat(len, join(@splat(a_vec_count, true), @splat(b_vec_count, false)));
+        const a_indices = count_up - cycle * @splat(len, @intCast(i32, b_vec_count));
+        const b_indices = shiftElementsUp(count_up - cycle * @splat(len, @intCast(i32, a_vec_count)), a_vec_count, 0);
+        break :blk @select(i32, select_mask, a_indices, ~b_indices);
+    };
+
+    return @shuffle(Child, a, b, indices);
+}
+
+/// The contents of braided is evenly split between vec_count vectors that are returned as an array. They "take turns",
+/// recieving one element from braided at a time.
+pub fn unbraid(comptime vec_count: usize, braided: anytype) [vec_count]Vector(vectorLength(@TypeOf(braided)) / vec_count, std.meta.Child(@TypeOf(braided))) {
+    const vec_len = vectorLength(@TypeOf(braided)) / vec_count;
+    const Child = std.meta.Child(@TypeOf(braided));
+
+    var out: [vec_count]Vector(vec_len, Child) = undefined;
+
+    comptime var i: usize = 0; // for doesn't work, apparently.
+    inline while (i < out.len) : (i += 1) {
+        const indices = comptime countUp(i32, vec_len) * @splat(vec_len, @intCast(i32, vec_count)) + @splat(vec_len, @intCast(i32, i));
+        out[i] = @shuffle(Child, braided, undefined, indices);
+    }
+
+    return out;
+}
+
+pub fn extract(vec: anytype, comptime first: VectorIndex(@TypeOf(vec)), comptime count: VectorCount(@TypeOf(vec))) Vector(count, std.meta.Child(@TypeOf(vec))) {
+    const Child = std.meta.Child(@TypeOf(vec));
+    const len = vectorLength(@TypeOf(vec));
+
+    std.debug.assert(@intCast(comptime_int, first) + @intCast(comptime_int, count) <= len);
+
+    return @shuffle(Child, vec, undefined, countUp(i32, count) + @splat(count, @intCast(i32, first)));
+}
+
+test "vector patterns" {
+    const base = Vector(4, u32){ 10, 20, 30, 40 };
+    const other_base = Vector(4, u32){ 55, 66, 77, 88 };
+
+    const small_bases = [5]Vector(2, u8){
+        Vector(2, u8){ 0, 1 },
+        Vector(2, u8){ 2, 3 },
+        Vector(2, u8){ 4, 5 },
+        Vector(2, u8){ 6, 7 },
+        Vector(2, u8){ 8, 9 },
+    };
+
+    try std.testing.expectEqual([6]u32{ 10, 20, 30, 40, 10, 20 }, repeat(6, base));
+    try std.testing.expectEqual([8]u32{ 10, 20, 30, 40, 55, 66, 77, 88 }, join(base, other_base));
+    try std.testing.expectEqual([2]u32{ 20, 30 }, extract(base, 1, 2));
+
+    if (!builtin.cpu.arch.isMIPS()) {
+        try std.testing.expectEqual([8]u32{ 10, 55, 20, 66, 30, 77, 40, 88 }, braid(.{ base, other_base }));
+
+        const small_braid = braid(small_bases);
+        try std.testing.expectEqual([10]u8{ 0, 2, 4, 6, 8, 1, 3, 5, 7, 9 }, small_braid);
+        try std.testing.expectEqual(small_bases, unbraid(small_bases.len, small_braid));
+    }
+}
+
+pub fn shiftElementsUp(vec: anytype, comptime amount: VectorCount(@TypeOf(vec)), shift_in: std.meta.Child(@TypeOf(vec))) @TypeOf(vec) {
```

Comment:
description? what is "up shifting"? shifting each element left? increasing each element with a shift?!

Reply:
It's shifting elements to higher indices. I'm not sure whether that'd be called a left- or a right-shift (it's towards the right of an array literal, but the bit positions of elements within the vector are technically left-shifted), so I went with up-shift. Thinking back, though, I must admit that that might've been a poor choice. Should I call it left or right?

Reply:
[I found this description for bi-endian](https://llvm.org/devmtg/2014-10/Slides/Schmidt-SupportingVectorProgramming.pdf),
so shifting up sounds reasonable if you add a comment (and ideally some ascii art explaining things for x86).
Do you know, if bi-endian architectures can change operation endianess at runtime?

I think MIPS fails, because it is used as big endian in CI.

Reply:
I'm staying away from raw memory access since it seems a bit risky in terms of platform-dependent undefined behavior.

Reply:
I would probably call it 'shifting left' and 'shifting right'.